### PR TITLE
Unlink unit charm dir symlink for application deployed before 2.8 bec…

### PR DESCRIPTION
**Pull request by ycliuhw**
### Checklist

 - [ ] ~Checked if it requires a [pylibjuju](https://github.com/juju/python-libjuju) change?~
 - [ ] ~Added [integration tests](https://github.com/juju/juju/tree/develop/tests) for the PR?~
 - [ ] ~Added or updated [doc.go](https://discourse.jujucharms.com/t/readme-in-packages/451) related to packages changed?~
 - [x] Do comments answer the question of why design decisions were made?

----

## Description of change

Unlink unit charm dir symlink for applications deployed before 2.8;

## QA steps

- Deploy k8s application on a 2.7 model;
- upgrade controller then model to 2.8
- check the model log to see if this error occur or not: cannot upgrade charm: rename /var/lib/juju/agents/unit-mariadb-k8s-0/charm/.juju-deploying /var/lib/juju/agents/unit-mariadb-k8s-0/charm/.juju-charm: no such file or directory

## Documentation changes

No

## Bug reference

https://bugs.launchpad.net/juju/+bug/1877025

### CODE REVIEW

**Comment by ycliuhw** [comment]
[https://bugs.launchpad.net/juju/+bug/1877025](https://trello.com/c/vKZTjkFo/1117-https-bugslaunchpadnet-juju-bug-1877025)

**Comment by ycliuhw** [comment]
Sure, I was about to add one today actually.

File: worker/caasoperator/caasoperator.go [hunk 1]
```diff
@@ -291,6 +290,23 @@ func (op *caasOperator) makeAgentSymlinks(unitTag names.UnitTag) error {
 		}
 	}
 
+	// Ensure legacy charm symlinks created before 2.8 getting unlinked.
```
Comment (hpidcock) [inline worker/caasoperator/caasoperator.go:293]: I'd refactor this like to keep the success flow to the end of the function.
Makes it easier for the next person to add to the function.
```
	// Ensure legacy charm symlinks created before 2.8 getting unlinked.
	unitCharmDir := filepath.Join(op.config.DataDir, "agents", unitTag.String(), "charm")
	isUnitCharmDirSymlink, err := jujusymlink.IsSymlink(unitCharmDir)
	if os.IsNotExist(errors.Cause(err)) || os.IsPermission(errors.Cause(err)) {
		// Ignore permission denied as this won't happen in production
		// but may happen in testing depending on setup of /tmp	
	} else if err != nil {
		return errors.Trace(err)
	} else if isUnitCharmDirSymlink {
		op.config.Logger.Debugf("unlinking legacy charm symlink for unit %q", unitTag)
		if err := os.Remove(unitCharmDir); err != nil {
			return errors.Trace(err)
		}
	}
```

**Comment by ycliuhw** [comment]
This PR has been merged to #11544, so close it now

**Pull request closed without merging**